### PR TITLE
docs: Add more ImplicitAllowedTypes docs

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Add more ImplicitAllowedTypes docs

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 1]
```diff
@@ -476,9 +476,34 @@ function evaluateLazySchema(value: LazyItem<TreeNodeSchema>): TreeNodeSchema {
 }
 
 /**
- * Types allowed in a field.
+ * Types of {@link TreeNode|TreeNodes} or {@link TreeLeafValue|TreeLeafValue} allowed at a location in a tree.
```
Comment (noencke) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:479]:
```suggestion
 * Types of {@link TreeNode|TreeNodes} or {@link TreeLeafValue|TreeLeafValues} allowed at a location in a tree.
```

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 2]
```diff
@@ -476,9 +476,34 @@ function evaluateLazySchema(value: LazyItem<TreeNodeSchema>): TreeNodeSchema {
...
  * @remarks
+ * Used by {@link TreeViewConfiguration} for the root and various kinds of {@link TreeNodeSchema} to specify their allowed child types.
+ *
+ * Use {@link SchemaFactory} to access leaf schema or declare new composite schema.
+ *
  * Implicitly treats a single type as an array of one type.
+ *
+ * Arrays of schema can be used to specify multiple types are allowed, which result in unions of those types in the Tree APIs.
+ *
+ * When saved into variables, avoid type erasing the details as doing so loses the compile time schema awareness of APIs derived from the types.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:489]:
```suggestion
 * When saved into variables, avoid type-erasing the details, as doing so loses the compile time schema awareness of APIs derived from the types.
```
An example of what not to do here might be useful.

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 3]
```diff
@@ -476,9 +476,34 @@ function evaluateLazySchema(value: LazyItem<TreeNodeSchema>): TreeNodeSchema {
...
+ * const config = new TreeViewConfiguration({ schema: childTypes });
+ * ```
+ *
+ * @example Forward reference
+ * ```typescript
+ * const sf = new SchemaFactory("myScope");
+ * class A extends sf.array("example", [() => B]) {}
+ * class B extends sf.array("Inner", sf.number) {}
+ * ```
  * @public
```
Comment (noencke) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:507]: Do you think this should be a @system type? Even if so, I guess it's technically too late to change it...

File: packages/dds/tree/src/simple-tree/schemaTypes.ts [hunk 4]
```diff
@@ -476,9 +476,34 @@ function evaluateLazySchema(value: LazyItem<TreeNodeSchema>): TreeNodeSchema {
...
+ * Use {@link SchemaFactory} to access leaf schema or declare new composite schema.
+ *
  * Implicitly treats a single type as an array of one type.
+ *
+ * Arrays of schema can be used to specify multiple types are allowed, which result in unions of those types in the Tree APIs.
+ *
+ * When saved into variables, avoid type erasing the details as doing so loses the compile time schema awareness of APIs derived from the types.
+ *
+ * When referring to types that are declared after the definition of the `ImplicitAllowedTypes`, the schema can be wrapped in a lambda to allow the forward reference.
+ * When constructing recursive schema see {@link ValidateRecursiveSchema} for details on how to structure the `ImplicitAllowedTypes` instances.
```
Comment (Josmithr) [inline packages/dds/tree/src/simple-tree/schemaTypes.ts:492]: Nit (I think this flows a bit better, personally)
```suggestion
 * See {@link ValidateRecursiveSchema} for details on how to structure the `ImplicitAllowedTypes` instances when constructing recursive schema.
```

**Review by Josmithr — APPROVED** [review]
These are great! Thank you for adding them!

File: packages/dds/tree/src/simple-tree/schemaTypes.ts (hunk 1 above)
Comment (Josmithr) [reply packages/dds/tree/src/simple-tree/schemaTypes.ts:479]: An alternative that might be simpler would be:
```suggestion
 * Types of {@link TreeNode}s or {@link TreeLeafValue}s allowed at a location in a tree.
```

But either way seems fine

File: packages/dds/tree/src/simple-tree/schemaTypes.ts (hunk 3 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/schemaTypes.ts:507]: This type is used as an input to tons of public APIs. Making it system would be strange and make it hard to write generic functions for anything schema related without referencing it. It really seems like one of the most key types in the public schema API surface used in every place we support polymorphism. It is non system on purpose (and also to late to change if we wanted)

**Pull request merged**